docs/configuration.md
# Installation and Configuration
## Required Settings
Required settings are stored as environment variables. These are the parameters responsible for authorization. Without these, you can only run commands that do not connect to the Matrix server.

| Variable            | Description                                                                                                   | Example                    |
| :------------------ | :------------------------------------------------------------------------------------------------------------ | :------------------------- |
| `MATRIX_HOMESERVER` | Matrix homeserver address.                                                                                    | `https://matrix.org`       |
| `MATRIX_USERNAME`   | Bot's account username. Create a user via Matrix Authentication Service (MAS): `docker exec matrix-auth mas-cli manage register-user USERNAME --password PASSWORD`                                                                                       | `@reminder-bot:matrix.org` |
| `MATRIX_PASSWORD`   | Bot's account password.                                                                                       | `mypassword`               |
| `MATRIX_TOKEN`      | Authentication token instead of a username and password. Generate a token via [Element Admin](https://github.com/element-hq/element-admin) or other service. | `mpt_mytoken`              |
| `MATRIX_RECOVERY`      | A recovery key that you enter only when you want to verify your account and then keep in a safe place. For more information, see the dedicated [Matrix account page](https://github.com/outarde/remindrix/tree/main/docs/matrix.md). | `recovery_key`              |

You can set variables in the [docker-compose.yml](https://github.com/outarde/remindrix/blob/main/docker/docker-compose.yml):
```yaml
environment:
  - MATRIX_HOMESERVER=https://matrix.org
  - MATRIX_TOKEN=mpt_mytoken
```
Or in the [.env file](https://github.com/outarde/remindrix/blob/main/docker/example.env) in the root of the Docker container folder:
```
MATRIX_HOMESERVER=https://matrix.org
MATRIX_USERNAME=@reminder-bot:matrix.org
MATRIX_PASSWORD=mypassword
```
---
The bot also has an optional variable `MATRIX_DEVICE` that specifies the device name. This name is displayed on the user's device list page and in the admin panel and does not affect the bot's use. Its default value is `reminder-bot-device`.

## Bot’s Optional Settings
Additional settings are stored in the `config.yml`/`config.yaml` file in a folder or volume that you have bound to the `/app/data/remindrix` folder inside the container. The repository has default [config.example.yml](https://github.com/outarde/remindrix/blob/main/docker/config.example.yml).

<details>
<summary>Interactive configurator</summary>

The interactive configurator walks you through a series of questions to create a settings file and saves it to disk. Run it via the command line:
```
docker run --rm -it -v ~/bot-data:/app/data/remindrix ghcr.io/outarde/remindrix:latest config-setup
```
The settings file will be saved in `~/bot-data` - in this case, in the folder in the user's home directory.

If you have already created a bot container, run the command where `remindrix` is the name of your container:

`docker exec -it remindrix config-setup`

Confirm that you want to create or overwrite a settings file.
</details>

| Field             | Description                                                                                                                                                                                          | Default        |
| :---------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------- |
| `lang`            | Language for bot commands and messages. See below for a list of available languages. The setting applies to all users who have not set a language.                                                                                           | `en`           |
| `remind_commands` | Aliases that override the standard bot invocation command. These are formatted as a list. The command for the selected language is available regardless of this variable.                            | `[ "remind" ]` |
| `list_commands`   | Same for the reminders list command. Does not have a localized version.                                                                                                                                                                | `[ "list" ]`   |
| `settings_commands`     | Same for the settings command. Does not have a localized version.                                                                                                                                                                | `[ "settings" ]`     |
| `on_command`      | Activate the bot only when a prefix (slash `/` or exclamation mark `!`) is presented at the beginning of a message. | `true`         |
| `on_command_group`      | The same as above but for rooms with more than two active or invited participants. | `true`         |
| `on_mention`      | Activation of the bot in chats with more than two active or invited participants only when it is mentioned: `@reminder-bot:matrix.org ...`.                                                          | `false`        |
| `quick_remind`      | If enabled, any text sent to the bot, except for other commands, will be treated as a new reminder. The setting does not apply to group rooms to prevent false positives.                                                         | `true`        |
| `remind_undated`      | Create reminders without requiring a date or time. Set the default morning time for today or tomorrow if today is too late. **Caution**: when enabled along with the `quick_remind` option, this will create reminders from any text sent to the bot.                                                          | `false`        |
| `send_reactions`      | The bot will send emoji reactions instead of success messages, but errors will remain in text format.                                                          | `true`        |
| `send_digits_reactions`      | The bot will send reactions in the form of a number of the longest duration before the reminder time. The Matrix doesn't allow two identical reactions to be sent to the same message, so numbers like 11 and 22 are replaced with the timer emoji ⏲️.                                                          | `true`        |
| `tz`         | Default time zone for all rooms on the server. Use [TZ identifier base](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones). | `Europe/Paris`        |
| `default_time`         | The time for the reminder when no time is specified. Please follow the format `%H:%M`, like `09:00`, `21:00`. The setting applies to all users who have not changed it.                                                                                                                                                                 | `10:00`        |
| `morning`         | The time that is considered morning.                | `09:00`        |
| `afternoon`       | The time that is considered afternoon.                                                                                                                                                               | `14:00`        |
| `evening`         | The time that is considered evening.                                                                                                                                                                 | `19:00`        |

> [!NOTE]
>  New versions may contain incompatible configuration changes. We'll reflect these in the releases notes.

## Language
### List of Available Languages
**From v0.4.0:**
- `en` English 🇬🇧,
- `de` German 🇩🇪,
- `fr` French 🇫🇷,
- `it` Italian 🇮🇹,
- `es` Spanish 🇪🇸,
- `sv` Swedish, aka IKEAish 🇸🇪,
- `pl` Polish 🇵🇱,
- `cs` Czech 🇨🇿,
- `fi` Finnish 🇫🇮,
- `ja` Japanese 🇯🇵,
- `zh` Chinese Simplified 🇨🇳,
- `ru` Russian 🇷🇺,
- `uk` Ukrainian 🇺🇦.

If you notice an incorrect translation or would like to request an other language, please [report it](https://github.com/outarde/remindrix/issues).
### Using a Custom Translation File
A custom translation file is a great way to add a language that isn't yet in the bot, or to customize an existing translation to suit your needs, for a themed homeserver or special occasion 🎃!

**Step one.** Create a `locales` folder in the folder already bound to `/app/data/remindrix`.

**Step two.** Create an `app.yml` file inside it. 

**Step three.** Add your translations, checking the keys from the [default localization file](https://github.com/outarde/remindrix/blob/main/locales/app.yml) and using the [standard language codes](https://en.wikipedia.org/wiki/List_of_ISO_639_language_codes).

Example:
```yaml
welcome.on_command: 
  en: >
    I'm a reminder bot. Creating a reminder is easy: 
    `!remind %{date} [at] 15:30 <reminder text>`. 
    I'll take care of the rest!
welcome.on_command_off: 
  en: >
    I'm a reminder bot. Creating a reminder is easy: 
    `%{date} [at] 15:30 <reminder text>`. 
    I'll take care of the rest!
reminder.command: 
  en: remind
reminder.saved: 
  en: ⏲️ Remind you on %{date} at %{hour}:%{min}
reminder.list: 
  en: >
    %{text} on %{date} at %{time}
reminder.new: 
  en: >
    🟢 Don't forget: %{text}
reminder.missed: 
  en: |
    ⚠️ You missed something: 
    %{sum}
reminder.error.month: 
  en: The thirteenth month!
reminder.error.time: 
  en: Oh, the times!
reminder.error.past-time: 
  en: Time is in the past, no reminder needed!
reminder.error.summer-time:
  en: This time does not exist in your time zone due to seasonal change.

# We want to use the default values for next keys, 
# so we don't include them in the file.
# months: 
#  en: ...
```
> [!TIP]
>  The `>` symbol means to remove all line breaks, the `|` symbol means to keep line breaks.

The bot will first search for a translation in your file, and then in the standard one.

If you've created a translation file that you'd like to share with the community, please make a pull request to the `/docs/locales` folder of this repository.

## What's Next
Read about the tools for working [with Matrix account](https://github.com/outarde/remindrix/tree/main/docs/matrix.md) (including device verification) or skip straight to the page about the intricacies [of using the bot](https://github.com/outarde/remindrix/tree/main/docs/usage.md).


src/config.rs
use tokio::fs;
use std::{
    fs as std_fs, 
    path::PathBuf
};
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use tracing::{info, warn, error};

/// Default language for bot messages.
pub const DEFAULT_LANG: &str = "en";
/// Default bot command for new reminder.
pub const DEFAULT_COMMAND: &str = "remind";
/// Default command for the list of reminders.
pub const DEFAULT_LIST_COMMAND: &str = "list";
/// Default timezone.
pub const DEFAULT_TZ: &str = "Europe/Paris";
/// Default bot settings command.
pub const DEFAULT_SETTINGS_COMMAND: &str = "settings";
/// Default times
pub const DEFAULT_TIME: &str = "10:00";
pub const DEFAULT_MORNING_TIME: &str = "09:00";
pub const DEFAULT_AFTERNOON_TIME: &str = "14:00";
pub const DEFAULT_EVENING_TIME: &str = "19:00";


/// Config for AppContext
#[derive(Debug, Deserialize)]
pub struct AppConfig {
    pub auth: MatrixConfig,
    pub recovery: Option<RecoveryConfig>,
    pub bot: BotConfig,
}

/// Config for matrix server authentication
#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase")]
pub struct MatrixConfig {
    pub homeserver: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
    pub token: Option<String>,
    pub device: Option<String>,
    pub recovery: Option<String>,
}

/// Recovery for account
#[derive(Debug, Serialize, Deserialize)]
pub struct RecoveryConfig {
    pub recovery_key: String,
    pub created_at: String,
}

/// Bot config
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub struct BotConfig {
    #[serde(default = "BotConfig::default_lang")]
    pub lang: String,
    #[serde(default = "BotConfig::default_remind_command")]
    pub remind_commands: Vec<String>,
    #[serde(default = "BotConfig::default_list_command")]
    pub list_commands: Vec<String>,
    #[serde(default = "BotConfig::default_settings_command")]
    pub settings_commands: Vec<String>,
    #[serde(default = "BotConfig::default_on_command")]
    pub on_command: bool,
    #[serde(default = "BotConfig::default_on_command_group")]
    pub on_command_group: bool,
    #[serde(default = "BotConfig::default_on_mention")]
    pub on_mention: bool,
    #[serde(default = "BotConfig::default_quick_remind")]
    pub quick_remind: bool,
    #[serde(default = "BotConfig::default_remind_undated")]
    pub remind_undated: bool,
    #[serde(default = "BotConfig::default_send_reactions")]
    pub send_reactions: bool,
    #[serde(default = "BotConfig::default_send_digits_reactions")]
    pub send_digits_reactions: bool,
    #[serde(default = "BotConfig::default_tz")]
    pub tz: String,
    #[serde(default = "BotConfig::default_time")]
    pub default_time: String,
    #[serde(default = "BotConfig::default_morning_time")]
    pub morning: String,
    #[serde(default = "BotConfig::default_afternoon_time")]
    pub afternoon: String,
    #[serde(default = "BotConfig::default_evening_time")]
    pub evening: String,

    // If the "strict" mode is defined in the config, then 
    // any settings will be saved under the bot ID, 
    // in fact all settings will be tied only to rooms.
    // The mode should be implemented only if we warn that when it is disabled,
    // you will need to delete all bot settings - 
    // otherwise user settings will always overlap old bot settings. 
    // Or another implementation of the mode: when it is disabled, we simply change the logic 
    // retrieving settings from the database for the priority of the user, not the bot. 
    // To do this, you don’t even have to change the database, but simply move the arguments 
    // user_id and bot_id are swapped in "non-strict" mode, so that the user ID 
    // was perceived as a bot ID with high priority.
    // That is, three modes: user settings priority, bot priority, bot only. 

    // #[serde(default = "BotConfig::default_strict_settings")]
    // pub strict_settings: bool,
}

impl BotConfig {
    fn default_lang() -> String {
        DEFAULT_LANG.into()
    }
    fn default_on_command() -> bool {
        true
    }
    fn default_on_command_group() -> bool {
        true
    }
    fn default_on_mention() -> bool {
        false
    }
    fn default_quick_remind() -> bool {
        false
    }
    fn default_remind_undated() -> bool {
        false
    }
    fn default_send_reactions() -> bool {
        true
    }
    fn default_send_digits_reactions() -> bool {
        true
    }
    fn default_remind_command() -> Vec<String> {
        vec![DEFAULT_COMMAND.to_string()]
        // let mut map = HashMap::new();

        // for serde_yaml_ng::Value type: HashMap<String, serde_yaml_ng::Value>
        // map.insert("remind".to_string(), vec![serde_yaml_ng::Value::String(DEFAULT_COMMAND.to_string())].into());

        // map.insert("remind".to_string(), vec![DEFAULT_COMMAND.to_string()]);
        // map.insert("list".to_string(), vec![DEFAULT_LIST_COMMAND.to_string()]);
        // map.insert("tz".to_string(), vec![DEFAULT_TIMEZONE_COMMAND.to_string()]);
        // map
    }
    fn default_list_command() -> Vec<String> {
        vec![DEFAULT_LIST_COMMAND.to_string()]
    }
    fn default_tz() -> String {
        DEFAULT_TZ.into()
    }
    fn default_settings_command() -> Vec<String> {
        vec![DEFAULT_SETTINGS_COMMAND.to_string()]
    }
    fn default_time() -> String {
        DEFAULT_TIME.into()
    }
    fn default_morning_time() -> String {
        DEFAULT_MORNING_TIME.into()
    }
    fn default_afternoon_time() -> String {
        DEFAULT_AFTERNOON_TIME.into()
    }
    fn default_evening_time() -> String {
        DEFAULT_EVENING_TIME.into()
    }
    fn _default_strict_settings() -> bool {
        false
    }

    /// Create config.yaml from config variable
    pub fn save_setup(&self, overwrite: bool, data_dir: &PathBuf, config: Option<BotConfig>) -> anyhow::Result<()> {
        let yaml_path = data_dir.join("config.yaml");
        if yaml_path.exists() && !overwrite {
            warn!("Configuration file already exists at {}", yaml_path.display());
            return Ok(());
        }

        let yaml_str = if let Some(c) = config {
            serde_yaml_ng::to_string(&c).context("Failed to serialize bot config")?
        }
        else {
            serde_yaml_ng::to_string(&self).context("Failed to serialize bot config")?
        };

        std_fs::write(&yaml_path, yaml_str)?;

        info!(file = %yaml_path.display(), "The data has been written to config.yaml.");

        Ok(())
    }
}

impl AppConfig {
    pub async fn load(data_dir: &PathBuf) -> Result<Self> {
        //dotenvy::dotenv().ok();

        // .env
        let auth: MatrixConfig = envy::prefixed("MATRIX_")
            .from_env()
            .map_err(|e| anyhow::anyhow!("Environment error: {}", e))?;

        // recovery.json
        // let data_dir = dirs::data_dir().context("No data_dir directory found")?.join(super::APP_FOLDER);
        let recovery_file = data_dir.join("recovery.json");

        let recovery = if recovery_file.exists() {
            let serialized = fs::read_to_string(&recovery_file).await
                .context("Error reading recovery.json")?;
            let data: RecoveryConfig = serde_json::from_str(&serialized)
                .context("File recovery.json has invalid JSON")?;
            Some(data)
        } else {
            None
        };

        // config.yaml or .env
        // yaml and yml files are supported
        let yaml_path = data_dir.join("config.yaml");
        let yml_path = data_dir.join("config.yml");
        let config_path = if yaml_path.exists() {
            Some(yaml_path)
        } else if yml_path.exists() {
            Some(yml_path)
        } else {
            None
        };
        
        let bot: BotConfig = if let Some(path) = config_path {
            let yaml_str = fs::read_to_string(&path).await
                .with_context(|| format!("Error reading {}", path.display()))?;
            
            match serde_yaml_ng::from_str::<BotConfig>(&yaml_str) {
                Ok(config) => {
                    info!(file = %path.display(), "Successfully loaded bot configuration");
                    config
                },
                Err(e) => {
                    error!(error = %e, "Failed to parse YAML config. Falling back to .env");
                    envy::prefixed("BOT_")
                        .from_env()
                        .map_err(|e| anyhow::anyhow!("Environment error: {}", e))?
                }
            }
        } else {
            warn!("No config.yaml or config.yml found, using .env variables");
            envy::prefixed("BOT_")
                .from_env()
                .map_err(|e| anyhow::anyhow!("Environment error: {}", e))?
        };

        // Check some values
        // Times
        check_times(vec![&bot.default_time, &bot.morning, &bot.afternoon, &bot.evening])?;
        tracing::info!("Default times are valid");

        super::settings_service::SettingsService::parse_optional_tz(Some(bot.tz.clone()))
            .ok_or(anyhow::anyhow!("The time zone in config.yaml has an invalid format: {}", bot.tz))?;
        tracing::info!("Default time zone is valid");

        Ok(Self { auth, recovery, bot })
    }
}

/// Just checks if string slices have valid time format via method in the reminder.rs module.
fn check_times(times: Vec<&str>) -> anyhow::Result<()> {
    for t in times {
        if !super::reminder::is_time_valid(t) {
            return Err(anyhow::anyhow!("Defaul time {} has an invalid format", t));
        }
    }
    Ok(())
}
